### PR TITLE
file out writer implementation

### PR DESCRIPTION
**Pull request by nasir-rabbani**
1. We can use `--log-output-dir` flag if we want to write the log and scan results in files
2. The log file will get appended while the scan output will be overwritten.
3. The format of the scan output file will be in sync with `-o` flag input
4. For now we will support output writing for every format except sarif (as sarif already writes the result to a file and this will require considering the file locations field in the sarif output)

### CODE REVIEW

File: docs/usage/command_line_mode.md [hunk 1]
```diff
@@ -339,5 +341,6 @@ Global Flags:
   -l, --log-level string     log level (debug, info, warn, error, panic, fatal) (default "info")
   -x, --log-type string      log output type (console, json) (default "console")
   -o, --output string        output type (human, json, yaml, xml, junit-xml, sarif, github-sarif) (default "human")
+      --output-dir string    directory path to over-write the default logs and output files ( default directory is user home directory)
```
Comment (cesar-rodriguez) [inline docs/usage/command_line_mode.md:344]: Should be "overwrite" without hyphen

File: test/e2e/help/golden/help_init.txt [hunk 2]
```diff
@@ -13,4 +13,5 @@ Global Flags:
   -l, --log-level string     log level (debug, info, warn, error, panic, fatal) (default "info")
   -x, --log-type string      log output type (console, json) (default "console")
   -o, --output string        output type (human, json, yaml, xml, junit-xml, sarif, github-sarif) (default "human")
+      --output-dir string    directory path to over-write the default logs and output files ( default directory is user's home directory)
```
Comment (cesar-rodriguez) [inline test/e2e/help/golden/help_init.txt]: Should be "overwrite" without hyphen

File: test/e2e/help/golden/help_scan.txt [hunk 3]
```diff
@@ -37,4 +37,5 @@ Global Flags:
   -l, --log-level string     log level (debug, info, warn, error, panic, fatal) (default "info")
   -x, --log-type string      log output type (console, json) (default "console")
   -o, --output string        output type (human, json, yaml, xml, junit-xml, sarif, github-sarif) (default "human")
+      --output-dir string    directory path to over-write the default logs and output files ( default directory is user's home directory)
```
Comment (cesar-rodriguez) [inline test/e2e/help/golden/help_scan.txt]: Should be "overwrite" without hyphen

File: test/e2e/help/golden/help_server.txt [hunk 4]
```diff
@@ -16,4 +16,5 @@ Global Flags:
   -l, --log-level string     log level (debug, info, warn, error, panic, fatal) (default "info")
   -x, --log-type string      log output type (console, json) (default "console")
   -o, --output string        output type (human, json, yaml, xml, junit-xml, sarif, github-sarif) (default "human")
+      --output-dir string    directory path to over-write the default logs and output files ( default directory is user's home directory)
```
Comment (cesar-rodriguez) [inline test/e2e/help/golden/help_server.txt]: Should be "overwrite" without hyphen

File: test/e2e/help/golden/help_unsupported_command.txt [hunk 5]
```diff
@@ -14,6 +14,7 @@ Flags:
   -l, --log-level string     log level (debug, info, warn, error, panic, fatal) (default "info")
   -x, --log-type string      log output type (console, json) (default "console")
   -o, --output string        output type (human, json, yaml, xml, junit-xml, sarif, github-sarif) (default "human")
+      --output-dir string    directory path to over-write the default logs and output files ( default directory is user's home directory)
```
Comment (cesar-rodriguez) [inline test/e2e/help/golden/help_unsupported_command.txt]: Should be "overwrite" without hyphen

File: test/e2e/help/golden/help_version.txt [hunk 6]
```diff
@@ -13,4 +13,5 @@ Global Flags:
   -l, --log-level string     log level (debug, info, warn, error, panic, fatal) (default "info")
   -x, --log-type string      log output type (console, json) (default "console")
   -o, --output string        output type (human, json, yaml, xml, junit-xml, sarif, github-sarif) (default "human")
+      --output-dir string    directory path to over-write the default logs and output files ( default directory is user's home directory)
```
Comment (cesar-rodriguez) [inline test/e2e/help/golden/help_version.txt]: Should be "overwrite" without hyphen

File: test/e2e/help/golden/no_command.txt [hunk 7]
```diff
@@ -19,6 +19,7 @@ Flags:
   -l, --log-level string     log level (debug, info, warn, error, panic, fatal) (default "info")
   -x, --log-type string      log output type (console, json) (default "console")
   -o, --output string        output type (human, json, yaml, xml, junit-xml, sarif, github-sarif) (default "human")
+      --output-dir string    directory path to over-write the default logs and output files ( default directory is user's home directory)
```
Comment (cesar-rodriguez) [inline test/e2e/help/golden/no_command.txt]: Should be "overwrite" without hyphen

**Review by cesar-rodriguez — CHANGES_REQUESTED** [review]
Few comments:
 
- I think the default behavior should be not to store the logs in a file just stdout like it currently works.
- The name of the flag should be "--log-output-dir" to be more specific 
- There's a typo where the word "over-write" should be "overwrite" 
- The functionality is not working for me for some reason. The file is showing empty

**Review by patilpankaj212 — COMMENTED** [review]
we should add an e2e test for this asserting that the log files are created in the provided directory.

File: pkg/writer/sarif.go [hunk 8]
```diff
@@ -103,7 +104,8 @@ func writeSarif(data interface{}, writer io.Writer, forGithub bool) error {
 	}
 
 	// print the report to anything that implements `io.Writer`
-	return report.PrettyWrite(writer)
+	// Sarif already writes the result in a file hence we are ignoring the file output writer
```
Comment (patilpankaj212) [inline pkg/writer/sarif.go:107]: so for sarif writer the output file won't be present in the provided log output directory?

**Comment by nasir-rabbani** [comment]
> we should add an e2e test for this asserting that the log files are created in the provided directory.

done

File: pkg/writer/writer.go [hunk 9]
```diff
@@ -28,13 +28,13 @@ var (
 )
 
 // Write method writes in the given format using the respective writer func
-func Write(format string, data interface{}, writer io.Writer) error {
+func Write(format string, data interface{}, writers []io.Writer) error {
```
[the remark this answers is not in the document]
Comment (nasir-rabbani) [reply pkg/writer/writer.go:31]: As discussed, we are considering the current approach.

File: pkg/writer/sarif.go (hunk 8 above)
Comment (nasir-rabbani) [reply pkg/writer/sarif.go]: added support for sarif as well.

**Review by cesar-rodriguez — DISMISSED** [review]

**Comment by cesar-rodriguez** [comment]
@nasir-rabbani there's something strange going on with the "security" check. Can you please try pulling the latest changes from master and pushing? Thanks

cc @bkizer-tenable

**Comment by nasir-rabbani** [comment]
> @nasir-rabbani there's something strange going on with the "security" check. Can you please try pulling the latest changes from master and pushing? Thanks
> 
> cc @bkizer-tenable

@cesar-rodriguez @bkizer-tenable 
The `Security` workflow is marked as required but the trigger event is set as `push` which is causing the PR checks to wait forever. 
Moreover, I tried to add the `pull-request` trigger event, but it failed as the secrets won't be accessible to PRs from a forked repository.
As a solution, I think instead of using the `Snyk Github Action` we should use the [Snyk Github App](https://github.com/marketplace/snyk)

**Review by Rchanger — APPROVED** [review]
LGTM.

**Pull request merged**